### ISSUE TITLE
[BUG] reinstall 1st node

### ISSUE THREAD

**Issue by tgazik** [harvester/harvester#2665]
**Describe the bug**
I tried to reinstall 1st node (installed with mode: create) and the node after reinstallation didn`t joined existing cluster.

**To Reproduce**
1. install 3 node cluster 
2. turn off 1st node (installed with mode: create), remove it from cluster via GUI - hosts - delete host, wipe all disks (with fdisk remove partititions and mkfs.ext4 /dev/sda) and reinstall this node (install mode: join) to join existing cluster with rest 2 nodes with new hostname.
3. reinstalled node is not able to join existing cluster (rancher bootstrap finished successfully, but node is not joined).

**Expected behavior**
Reinstalled node should join existing cluster with rest 2 nodes.

**Support bundle**
can provide if needed

**Environment**
 - Harvester ISO version: v1.0.3
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): tried on 2 different environments with different HW

**Additional context**
When i tried to reinstall 2nd or 3rd node (installation mode: join) they join the cluster successfully after reinstallation. The only problem is with the 1st node.

**Comment by w13915984028** [comment]
@tgazik Please add the support bundle file, thanks.
 https://docs.harvesterhci.io/v1.0/troubleshooting/harvester/#generate-a-support-bundle
 
As a workaround, could you please try to install the first node again, with a different hostname and host IP ?  The failure may be related to reusing the same hostname and IP, but there are potential conflict remainings, we are investigating.

**Comment by tgazik** [comment]
Tried to replicate with different hostname and IP, but the result is same, node didn`t joined the cluster.
i can see in journalctl -b -u rancherd:
```
2022-08-15T11:02:22Z" level=info msg="Bootstrapping Rancher (v2.6.4-harvester3/v1.22.12+rke2r1)"
"2022-08-15T11:02:23Z" level=info msg="Writing plan file to /var/lib/rancher/rancherd/plan/plan.json"
"2022-08-15T11:02:23Z" level=info msg="Applying plan with checksum "
"2022-08-15T11:02:23Z" level=info msg="No image provided, creating empty working directory /var/lib/rancher/rancherd/plan/work/>
"2022-08-15T11:02:23Z" level=info msg="Running command: /usr/bin/env [sh /var/lib/rancher/rancherd/install.sh]"
"2022-08-15T11:02:23Z" level=info msg="[stdout]: [INFO]  Using default agent configuration directory /etc/rancher/agent"
"2022-08-15T11:02:23Z" level=info msg="[stdout]: [INFO]  Using default agent var directory /var/lib/rancher/agent"
"2022-08-15T11:02:23Z" level=info msg="[stderr]: [WARN]  /usr/local is read-only or a mount point; installing to /opt/rancher-s>
"2022-08-15T11:02:25Z" level=info msg="[stdout]: [INFO]  Determined CA is necessary to connect to Rancher"
"2022-08-15T11:02:25Z" level=info msg="[stdout]: [INFO]  Successfully downloaded CA certificate"
"2022-08-15T11:02:26Z" level=info msg="[stdout]: [INFO]  Value from https://192.168.44.10:443/cacerts is an x509 certificate"
"2022-08-15T11:02:26Z" level=info msg="[stdout]: [INFO]  Successfully tested Rancher connection"
"2022-08-15T11:02:26Z" level=info msg="[stdout]: [INFO]  Downloading rancher-system-agent from https://192.168.44.10:443/assets>
"2022-08-15T11:02:27Z" level=info msg="[stdout]: [INFO]  Successfully downloaded the rancher-system-agent binary."
"2022-08-15T11:02:27Z" level=info msg="[stdout]: [INFO]  Generating Cattle ID"
"2022-08-15T11:02:27Z" level=info msg="[stdout]: [INFO]  Successfully downloaded Rancher connection information"
"2022-08-15T11:02:27Z" level=info msg="[stdout]: [INFO]  systemd: Creating service file"
"2022-08-15T11:02:27Z" level=info msg="[stdout]: [INFO]  Creating environment file /etc/systemd/system/rancher-system-agent.env"
"2022-08-15T11:02:28Z" level=info msg="[stdout]: [INFO]  Enabling rancher-system-agent.service"
"2022-08-15T11:02:28Z" level=info msg="[stderr]: Created symlink /etc/systemd/system/multi-user.target.wants/rancher-system-age>
"2022-08-15T11:02:28Z" level=info msg="[stdout]: [INFO]  Starting/restarting rancher-system-agent.service"
"2022-08-15T11:02:28Z" level=info msg="No image provided, creating empty working directory /var/lib/rancher/rancherd/plan/work/>
"2022-08-15T11:02:28Z" level=info msg="Running command: /usr/bin/rancherd [probe]"
"2022-08-15T11:02:29Z" level=info msg="[stderr]: time=\"2022-08-15T11:02:29Z\" level=info msg=\"Running probes defined in /var/>
"2022-08-15T11:02:30Z" level=info msg="[stderr]: time=\"2022-08-15T11:02:30Z\" level=info msg=\"Probe [kubelet] is unhealthy\""
```
no other lines in that log

support bundle before 1st node was reinstalled:
[github_issue-2665_before_1stnode-reinstall-supportbundle_e00fe9f4-4d91-4d66-be5c-76a5add6e1e1_2022-08-15T10-32-39Z.zip](https://github.com/harvester/harvester/files/9336833/github_issue-2665_before_1stnode-reinstall-supportbundle_e00fe9f4-4d91-4d66-be5c-76a5add6e1e1_2022-08-15T10-32-39Z.zip)

support bundle after 1st node was reinstalled:
[github_issue-2665_after_1stnode-reinstall-supportbundle_e00fe9f4-4d91-4d66-be5c-76a5add6e1e1_2022-08-15T11-12-34Z.zip](https://github.com/harvester/harvester/files/9336836/github_issue-2665_after_1stnode-reinstall-supportbundle_e00fe9f4-4d91-4d66-be5c-76a5add6e1e1_2022-08-15T11-12-34Z.zip)

**Comment by w13915984028** [comment]
@tgazik If the hanging node is still there, could you please help grab more information, which is not in the support-bundle. Thanks.

```
sudo -i

uname -a
ip link
ip addr
ps aux
systemctl status rancherd
systemctl status rke2-agent.service
cat /var/lib/rancher/rke2/agent/logs/kubelet.log
```

**Comment by tgazik** [comment]
@w13915984028 i already reinstalled the env, but i did the test again. See the outputs from requested commands
[requested_commands.txt](https://github.com/harvester/harvester/files/9337693/requested_commands.txt)

**Comment by w13915984028** [comment]
@tgazik Thanks for your fast retest. The attached information is very helpful. Will update when we find the root cause.

**Comment by tjjh89017** [comment]
@tgazik I will take a look.
Do you change first node IP in reinstallation process?

**Comment by tjjh89017** [comment]
note:
rancher-system-agent debug log
keep showing log below
```
Aug 17 05:55:16 node4 rancher-system-agent[3652]: time="2022-08-17T05:55:16Z" level=debug msg="[K8s] Processing secret custom-25229e2c3ace-machine-plan in namespace fleet-local at generation 0 with resource version 746934"
```

**Comment by tgazik** [comment]
@tjjh89017 
> @tgazik I will take a look. Do you change first node IP in reinstallation process?

Yes i changed the IP and also the hostname.

> note: rancher-system-agent debug log keep showing log below
> 
> ```
> Aug 17 05:55:16 node4 rancher-system-agent[3652]: time="2022-08-17T05:55:16Z" level=debug msg="[K8s] Processing secret custom-25229e2c3ace-machine-plan in namespace fleet-local at generation 0 with resource version 746934"
> ```
The host is still not joined into the cluster after 2 days. I keep the cluster installed.

**Comment by tjjh89017** [comment]
> The host is still not joined into the cluster after 2 days. I keep the cluster installed.

We are working with Rancher team to debug.
If possible, could you keep that env?
If rancher team needs more details.
Thanks a lot

**Comment by tgazik** [comment]
> > The host is still not joined into the cluster after 2 days. I keep the cluster installed.
> 
> We are working with Rancher team to debug. If possible, could you keep that env? If rancher team needs more details. Thanks a lot

Sure i will keep it, no problem. Thanks for update

**Comment by tjjh89017** [comment]
@tgazik we are still working on it.
and it seems it's easy to reproduce.
So if you need the cluster, you could reinstall it.
Thanks a lot

**Comment by Oats87** [comment]
@tjjh89017 is there a reproduction environment the Rancher team (I) can get access to? 

Unfortunately we are not operating with a lot of information here and I think I need to reproduce this in order to test it. 

What is the easiest way for me to reproduce this in my own environment?

@tgazik If you still have your environment around, would you be able to dump the logs of the `rke2-server` systemd service on your first controlplane node? My guess is that the `rke2-server` is unable to join the rest of the nodes, or the Rancher v2prov planner is not properly delivering a new plan to the node.

**Comment by Oats87** [comment]
This may be a duplicate of https://github.com/harvester/harvester/issues/2191

**Comment by tjjh89017** [comment]
> What is the easiest way for me to reproduce this in my own environment?

@Oats87 
I used laptop and create 3 VM to install harvster and reinstall first node and join back.
but I think the problem is a duplicate of 2191.
I used workaround to deal with it and need your advice to upgrade RKE2
https://github.com/rancher/rancher/issues/38640#issuecomment-1240456717

**Comment by FrankYang0529** [comment]
This issue can be fixed https://github.com/harvester/harvester-installer/pull/344 and https://github.com/rancher/rancherd/pull/25. Following are my test steps with [vagrant-pxe-harvester](https://github.com/harvester/ipxe-examples).

Case 1: Remove the second node
1. Setup a 3-node cluster (harvester-node-0, harvester-node-1, harvester-node-2).
2. After all 3-node cluster is installed, please run `kubectl delete node harvester-node-1` to remove the node CR.
3. After removing the node CR, please run `vagrant destroy harvester-node-1` to remove the node VM.
4. Run `vagrant up harvester-node-1` to start the node again. It should join the cluster and the status will be `Ready`.

Case 2: Remove the first node
1. Follow Case 1.
2. Run `kubectl delete node harvester-node-0` to remove the node CR.
3. After removing the node CR, please run `vagrant destroy harvester-node-0` to remove the node VM.
4. Run `vagrant ssh pxe_server` to update some content.
```
# copy config-join-1.yaml to config-join-0.yaml
cp /var/www/harvester/config-join-1.yaml /var/www/harvester/config-join-0.yaml

# update os.hostname to harvester-node-0 in config-join-0.yaml
vim /var/www/harvester/config-join-0.yaml

# change owner of config-join-0.yaml
chown www-data /var/www/harvester/config-join-0.yaml

# update 02:00:00:0d:62:e2. In the last line, change config-create.yaml to config-join-0.yaml
vim /var/www/harvester/02:00:00:0d:62:e2

# restart nginx
systemctl restart nginx.service
```
5. Run `vagrant up harvester-node-0` to start the node again. It should join the cluster and the status will be `Ready`.

**Comment by bk201** [comment]
Thanks to @FrankYang0529 for the verification!

**Comment by w13915984028** [comment]
@FrankYang0529  Great!

Please help check if #2470 is also caused by this bug, thanks.

**Comment by TachunLin** [comment]
### Test Plan 1: Reinstall management node and agent node in a upgraded cluster

### Result
Verified after upgrading from v1.0.3 to v1.1.0 master release, we **can rejoin** the management node and agent node back correctly. 

1. Successfully re-join the management node after upgrade 
  ![image](https://user-images.githubusercontent.com/29251855/196159969-9c8acb11-b9fe-4501-94d5-74545579ef4d.png)

2. Successfully re-join the agent node after upgrade 
  ![image](https://user-images.githubusercontent.com/29251855/196164183-20c4498a-c4c6-4033-a527-d3c8ff84f8aa.png)


### Test Information
* Test Environment: 4 nodes harvester on local kvm machine
* Harvester version: `master-0a9538a1-head` (10/14)

### Verify Steps
1. Create a 4-node v1.0.3 cluster.
2. Upgrade the master branch:
 - Check the spec content in `provisioning.cattle.io/v1/clusters -> fleet-local`
![image](https://user-images.githubusercontent.com/29251855/196139161-7b6e6e84-692d-4f4f-a978-62fc50f64f06.png)
 - Check the iface content in `helm.cattle.io/v1/helmchartconfigs -> rke2-canal`
![image](https://user-images.githubusercontent.com/29251855/196141139-e4ff668e-287a-4722-865d-a7c0f145c862.png)
  ```
   spec:                                                                                                                                                                                      │
  │   valuesContent: |-                                                                                                                                                                        │
  │     flannel:                                                                                                                                                                               │
  │       iface: "" 
  ```

3. Remove the agent node and 1 management node.
   - Remove agent node (node 4)
  ![image](https://user-images.githubusercontent.com/29251855/196138324-83b64d50-0236-4110-86c2-551ae046a406.png)
  ![image](https://user-images.githubusercontent.com/29251855/196138906-e77de8f0-d61d-4a91-82b2-e36cbbb6462a.png)
  
  - Remove management node (node 3)
  ![image](https://user-images.githubusercontent.com/29251855/196139707-ce8cc444-21fb-4bba-a040-9b09e0ae7fa9.png)
  ![image](https://user-images.githubusercontent.com/29251855/196139906-ff92320f-f5f7-4fb5-af20-5cf0f3fe36b9.png)
  ![image](https://user-images.githubusercontent.com/29251855/196147230-7c0b1835-c219-41b2-b0b8-32bae9962f27.png)

4. After the node is removed, provision a new node.
5. Check the node can join the cluster and can be promoted as a management node.
6. After we have 3 management nodes, provision a new node and check if it can join the cluster.

**Comment by TachunLin** [comment]
### Test Plan 2: Reinstall management node and agent node in a new Harvester cluster

### Result
Verified after upgrading from v1.0.3 to `v1.1.0-rc3`,we **can rejoin** the management node and agent node back correctly. 

1. Successfully re-join the management node after upgrade 
![image](https://user-images.githubusercontent.com/29251855/196353489-ea323d20-4028-411a-ba32-5fbb7ae4ae2f.png)
![image](https://user-images.githubusercontent.com/29251855/196353521-6882c3e4-f374-4c9f-a190-8b5929b8e9e5.png)

2. Successfully re-join the agent node after upgrade 
![image](https://user-images.githubusercontent.com/29251855/196358018-f9b7f72c-cfa4-4e6f-a25e-a8818e1f491b.png)
![image](https://user-images.githubusercontent.com/29251855/196358073-29b879be-1e0c-465e-9327-6430111d4436.png)

### Test Information
* Test Environment: 4 nodes harvester on local kvm machine
* Harvester version: `v1.1.0-rc3`

### Verify Steps
1. Create a 4-node v1.0.3 cluster.
2. Upgrade to the v1.1.0-rc3
3. Remove the agent node and the first management node 
   - Remove agent node (node 4)
   ![image](https://user-images.githubusercontent.com/29251855/196345973-36d62a3d-e2a5-4261-90a6-4d7e89425726.png)
   ![image](https://user-images.githubusercontent.com/29251855/196346046-76f1ceb3-a5a2-4988-bb3f-3dc090383199.png)
   ![image](https://user-images.githubusercontent.com/29251855/196346128-5e69b595-241a-4c15-b6ac-2f967e4561f4.png)

  - Remove management node (node 1)
   ![image](https://user-images.githubusercontent.com/29251855/196347345-1ff6de57-8870-4592-ba36-1039be6a9e09.png)
   ![image](https://user-images.githubusercontent.com/29251855/196347501-417a40a5-a669-4d21-b94e-d9f546737085.png)

4. After the node is removed, provision a new node.
5. Check the node can join the cluster and can be promoted as a management node. (node 1) 
6. After we have 3 management nodes, provision a new node and check if it can join the cluster. (node 4)

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.1.0-rc3` and `master-0a9538a1-head` (10/14). Close this issue.

### Result
Now we can rejoin the 1st management node and agent node in the following scenarios as stated above
1. Harvester upgrade from v1.0.3 to v1.1.0
2. Fresh install v1.1.0 Harvester
